4ian/GDevelop
# [Platformer] Remove the rounding property

## Solution
The pixel by pixel loop to set the object on the floor is replaced by a method that finds the floor position from platforms hit-boxes.

On a flat floor without pixel rounding, there was [0;1[ pixel interstice between the object and the platform. Now, if the platform is pixel align, the object will be too which make the rounding property useless.

The object will always land at the exact same Y for a give X on a platform. This will make issues easier to reproduce and tests stronger.

This PR seems to fix these issues
* https://github.com/4ian/GDevelop/issues/3013
* https://github.com/4ian/GDevelop/issues/897

Even if this issue is not a bug, it kind of fix it for flat floors as the collision will always be false with excluding edge and true if including them.
* https://github.com/4ian/GDevelop/issues/1920

It doesn't fix
* https://github.com/4ian/GDevelop/issues/2562

## Automatic tests
The tests stay the same apart from 2 places where odd one frame state doesn't appear anymore.
I compared the coordinates on the test with a slope, the values have less than one pixel difference.
There is a skipped test that passes even though it was not the objective. It might solve https://github.com/4ian/GDevelop/issues/2562

## Manual tests
I manually check slopes on the following example. I had to raise the 1 pixel margin to 1.9 to avoid the object to decelerate more often at a slope. Which is expected because the slope are followed precisely whereas the 1pixel search put it up to one pixel too high. After that, the behavior looks the same than with the 115. I opened an issue about that: https://github.com/4ian/GDevelop/issues/3013
The platformer behavior doesn't make any difference between a small cliff and a slope. This will need a rework too to avoid the deceleration when the slope changes.
https://www.dropbox.com/s/4pb6kiav6qeilck/PlatformTest.zip?dl=1

## Custom build
* If you want to try this preview version, please backup your game file before open it.
  https://www.dropbox.com/s/om9og4qai7sea1l/gdevelop-platformer-preview-beta121-2021-11-02.7z?dl=1

## Documentation

### X Search

![SlopeFollowingRequestedDeltaX](https://user-images.githubusercontent.com/2611977/140380544-569e561b-610b-4fe7-81f7-a78ca0a32a1e.png)

![SlopeFollowingClimbFactor](https://user-images.githubusercontent.com/2611977/140380583-4799312f-9339-42d9-94a6-9fb9db3b1053.png)

### Y Search Context

![SlopeFollowingContext](https://user-images.githubusercontent.com/2611977/140319740-9b959997-9c9e-4278-b973-bf0929c0900c.png)

### Y Search Result

![SlopeFollowingResult](https://user-images.githubusercontent.com/2611977/138004087-dfdb376e-4fad-457e-afc3-46c6d39a9ce8.png)
